compose/config/config.py

```diff
@@ -320,6 +321,32 @@ def find_candidates_in_parent_dirs(filenames, path):
     return (candidates, path)
 
 
+def check_swarm_only_config(service_dicts):
+    warning_template = (
+        "Some services ({services}) use the '{key}' key, which will be ignored. "
+        "Compose does not support '{key}' configuration - use "
+        "`docker stack deploy` to deploy to a swarm."
+    )
+
+    services_using_deploy = [s for s in service_dicts if s.get('deploy')]
+    services_using_credential_spec = [s for s in service_dicts if s.get('credential_spec')]
+    if services_using_deploy:
+        log.warn(
+            warning_template.format(
+                services=", ".join(sorted(s['name'] for s in services_using_deploy)),
+                key='deploy'
+            )
+        )
+
+    if services_using_credential_spec:
+        log.warn(
+            warning_template.format(
+                services=", ".join(sorted(s['name'] for s in services_using_credential_spec)),
+                key='credential_spec'
+            )
+        )
```

Comment:
Maybe split this into a new function?

```python

def check_swarm_only_key(service_dicts, key):
    ...
```

Then `check_swarm_only_config` could be:

```python
def check_swarm_only_config(service_dicts):
    check_swarm_only_key(service_dicts, 'deploy')
    check_swarm_only_key(service_dicts, 'credential_spec')
```